### PR TITLE
[27873] New Sidebar in mobile view and some mobile issues fixed

### PR DESCRIPTION
**Pull request by ingamai**
Issues of new sidebar fixed:
- [x] new sidebar slides in from the left side of the screen 
- [x] other menus are not overlapping the new sidebar anymore
- [x] top bar is sticky to the top (does not scroll with the content anymore)
- [x] icon color changes when hovered (in desktop view) or clicked in mobile view

Other mobile issues fixed:
- [x] roadmap view now fits the screen size
- [x] bug at wiki description in editing mode fixed

To-Do's
- [x] bug: sidebar is not scrollable until the end of the content
- [x] only transform the hamburger icon into a closing icon in the mobile view
- [ ] check breakpoints

https://community.openproject.com/projects/openproject/work_packages/27873/activity

### CODE REVIEW

**Comment by wielinde** [comment]
When I close the main menu, the hamburger does not have the right background color: 
image

**Comment by wielinde** [comment]
@ingamai Scrolling within the menu does scroll below the menu and not the menu. 

scroll-broken

**Comment by wielinde** [comment]
@lindenthal (@ingamai) Not sure if that is what you specified but I find it confusing to have the x-icon all the time open. I believe on Desktop the hamburger should only convert into an x when the mouse hovers above it. On mobile it could stay as it currently is in this PR, as the menu does not stay open.

image

**Comment by wielinde** [comment]
@ingamai the roadmap already looks much better on mobile. However, the progressbar still needs some changes, I believe:

image

**Comment by ingamai** [comment]
@wielinde The progress bar should now be adapted to the screen size, the scrolling content behind the open sidebar fixed and hopefully the problem with the false color is also fixed. This should be tested again later, as discussed.

**Comment by ingamai** [comment]
@wielinde The hamburger icon should now only transform into a closing icon in mobile view or on desktop if the menu is expanded and you hover over the hamburger icon. 
I don't know why, but the solution with the mixin didn't really work for me. So I changed the content of the element.

**Comment by wielinde** [comment]
@ingamai Great! Scrolling works now and the close icon, too. I can not find the color issue anymore either.

There is one thing that is not optimal but good enough for me at the moment: When you close menu with a click right from the main menu and that click hits a button then this button gets executed. That is clearly not what a user would expect. To prevent this, there must be some kind of (semi)transparent layer, that covers everything below the menu and "swallows" the click event. 

closing-outside-triggers-action-below

Then I found some other minor, probably older, padding issues, that would be nice if they were fixed:

In full view of a work package:
task__see_a_screenshot___11____demo_project___openproject

In _My profile_
profile___my_account___openproject

In "show" of a wiki page:
wiki___user_s_guide___openproject

**Comment by wielinde** [comment]
@ingamai Oh, and yes: Roadmap looks great!

**Comment by HDinger** [comment]
@wielinde shall the last issues you mentioned be fixed in this PR? Or shall we open a new ticket for this? I am asking because we could probably find a lot more issues with the mobile view and have to make a cut somewhere..

**Comment by wielinde** [comment]
@HDinger @ingamai First of all, I didn't review the code yet, but simply clicked through and reported what I found. 

I agree that it is better to quickly review and merge whatever improvement. So if you want to review it, please go ahead.

Yes, we could create separate work packages for each issue. However, at least the one with the padding above the WP full view is a regression. That was introduced in this PR and it is a pretty prominent place, considering that WPs is what you most look at.

**Comment by HDinger** [comment]
Ok 👍 
@ingamai I suggest the following: You fix everything Wieland mentioned that is directly related to the mobile menu navigation (or issues created by this PR).  This would be:

* The padding above the WP
* Handle the click outside the menu

For all other issues I will open a separate ticket (the wiki page, the profile page, further bugs on the roadmap etc.). https://community.openproject.com/projects/openproject/work_packages/27946/activity

**Comment by HDinger** [comment]
The scrolling bug seems to be fixed on all pages except the WP full view. There the menu is not scrollable but the content.

File: app/assets/stylesheets/content/_advanced_filters.sass [hunk 1]
```diff
@@ -168,7 +168,12 @@
...
     display: none
 
-
-
 fieldset#date-range p
   margin: 2px 0 2px 0
+
+@media only screen and (max-width: 42.4375em)
+  .advanced-filters--filters
+    width: 300px
```
Comment (HDinger) [inline app/assets/stylesheets/content/_advanced_filters.sass:176]: Why this change? The limit is really strict. This breaks the filter layout when there is a filter with two select fields. If the breakpoint is necessary please use the px value instead. I know that it is written this way in the style guide, but all other breakpoints listen to 679px. We should take care that they match at all cases. 
bildschirmfoto 2018-06-29 um 14 34 55

File: app/assets/stylesheets/content/_forms_mobile.sass [hunk 2]
```diff
@@ -80,3 +80,6 @@
     input[type="tel"],
     input[type="color"]
       font-size: 16px
+
+    textarea, iframe
```
Comment (HDinger) [inline app/assets/stylesheets/content/_forms_mobile.sass:84]: What is this needed for? It adds a lot of space at the top of **all** textareas. We should be careful when doing sth like this, because it could cause side effects we are not aware of. 

For example:
bildschirmfoto 2018-06-29 um 14 37 11

File: app/assets/stylesheets/content/_progress_bar.sass [hunk 3]
```diff
@@ -47,3 +47,17 @@
 
 .progress-bar-legend
   font-size: 80%
+
+// Prevent that the progress bar is cut off on mobile screens
+@include breakpoint(830px down)
```
Comment (HDinger) [inline app/assets/stylesheets/content/_progress_bar.sass:52]: 830px is a strange breakpoint as it is only used on landscape mobile devices. Please use either an implementation that works on mobile and desktop OR use a more common breakpoint.

File: app/assets/stylesheets/layout/_base_mobile.sass [hunk 4]
```diff
@@ -30,21 +30,22 @@
...
-  body
-    overflow: auto !important
-    overflow-y: scroll /* has to be scroll, not auto */
+  body,
+  #main,
+  #content,
+  #content-wrapper
     -webkit-overflow-scrolling: touch
+    overflow: visible
+  .lock-screen
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_base_mobile.sass:39]: In my eyes `lock-screen` is a bit confusing name (I thought of a screen to lock something). Maybe sth like `-locked-screen` would be better. But this could be very subjective, so I leave it up to you if you change this.

File: app/assets/stylesheets/layout/_base_mobile.sass [hunk 5]
```diff
@@ -74,3 +78,11 @@
 @include breakpoint(680px up)
   .hidden-for-desktop
     display: none
+
+// For mobile and tablet size
+@media (orientation: landscape) and (max-width: 830px) and (max-height: 500px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_base_mobile.sass:83]: Same as above (magic numbers). 
Why is this breakpoint so restrictive? Wouldn't `@media screen and (max-width:680px)` be sufficient?

File: app/assets/stylesheets/layout/_top_menu_mobile.sass [hunk 6]
```diff
@@ -108,7 +109,7 @@
         // Viewport minus #account-nav-right when logged out, which is
         // 165px, as this is the space for the long login text label: "Connexion" (French).
         // 36px is the padding around the project name.
-        max-width: calc(100vW - 165px - 36px - #{$hamburger-width + $hamburger-right})
+        max-width: calc(100vw - 160px - 22px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_top_menu_mobile.sass:112]: Please do not change numbers without updating the comment (when there is an explicit one explaining them). Where is the 22 coming from?
Unfortunately this does not work.  When the projects title is long there is an overlapping. 
bildschirmfoto 2018-06-29 um 15 01 34

File: app/assets/stylesheets/layout/_top_menu_mobile.sass [hunk 7]
```diff
@@ -133,3 +134,19 @@
     .project-menu-autocomplete--results
       border-bottom: 2px solid $header-drop-down-border-color !important
 
+  // Change hamburger icon into closing icon, if menu is expanded
+  #main-menu-toggle
+    &:not(.-collapsed)
+      i::before
+        content: "\f12c"
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_top_menu_mobile.sass:141]: Same as above

File: app/assets/stylesheets/layout/_top_menu.sass [hunk 8]
```diff
@@ -330,6 +330,13 @@ input.top-menu-search--input
     @include varprop(font-size, header-item-font-size)
     text-decoration: none
     padding: 0 15px
+  &:not(.-collapsed)
+    @include varprop(background, header-item-bg-hover-color)
+  &:hover
+    @include varprop(background, header-item-bg-hover-color)
+    &:not(.-collapsed)
+      i::before
+        content: "\f12c"
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_top_menu.sass:339]: The icon code will change when we update the icon font the next time. Please use `@include icon-mixin-close`

File: app/assets/stylesheets/layout/_top_menu_mobile.sass [hunk 9]
```diff
@@ -133,3 +134,19 @@
...
+
+// Workaround on tablet and mobile to prevent an icon having the wrong color after touch
+// After the touch event the focus is still staying on this element (unlike a mouse hover)
+@media (max-device-width: 1400px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_top_menu_mobile.sass:145]: Same as above

File: app/assets/stylesheets/layout/_work_package_mobile.sass [hunk 10]
```diff
@@ -93,7 +93,10 @@
             margin: 0.75rem 0 2.5rem 0
 
           .work-packages--panel-inner
-            padding-left: 0
+            padding-right: 0
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_work_package_mobile.sass:96]: Why did you do this? Because of this the content sticks directly to screen border..
bildschirmfoto 2018-06-29 um 15 14 15

File: app/assets/stylesheets/layout/_work_package_table.sass [hunk 11]
```diff
@@ -48,15 +48,15 @@
 // This makes use of more specific rules overwriting less specific ones.
 // Per default, the height is always 100%
 .work-packages-page--ui-view
-  height: 100%
+  min-height: calc(100vh - 55px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_work_package_table.sass:51]: Please use the `header-height` variable

File: app/assets/stylesheets/layout/_work_package_table.sass [hunk 12]
```diff
@@ -48,15 +48,15 @@
...
 
 .work-packages-list-view--container
   // Set  padding on left side
   padding-left: 20px
   // Flexbox for the toolbar, filters and work package split view
   display: flex
   flex-direction: column
-  height: 100%
+  min-height: calc(100vh - 55px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_work_package_table.sass:59]: Same

File: app/assets/stylesheets/layout/_work_package_table.sass [hunk 13]
```diff
@@ -192,13 +192,28 @@
     .form--field-container
       max-width: 400px
 
-// Workaround to make the toolbar menu accessible even on very small screens
-@media screen and (max-height: 25em)
+@media only screen and (max-width: 42.4375em)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_work_package_table.sass:195]: Please combine these two breakpoints (42.4375em = 679px)

File: app/assets/stylesheets/layout/_work_package_table.sass [hunk 14]
```diff
@@ -192,13 +192,28 @@
...
+    // Set  padding on left side
+    padding-left: 0
+
+    > .toolbar-container
+      padding-right: 0
+
+@media screen and (max-width:680px)
+  .work-packages-page--ui-view,
+  .work-packages-list-view--container
+    min-height: calc(100vh - 55px)
```
Comment (HDinger) [inline app/assets/stylesheets/layout/_work_package_table.sass:206]: Variable again

File: app/assets/stylesheets/openproject/_legacy.sass [hunk 15]
```diff
@@ -593,7 +593,8 @@ h4.comment
     // Neccessary so that button and select box have the same height
     line-height: 0.6rem
 
-@include breakpoint(680px down)
+// Breakpoint for big mobile screens in landscape mode
+@include breakpoint(830px down)
```
Comment (HDinger) [inline app/assets/stylesheets/openproject/_legacy.sass:597]: Is this really necessary? Same issue as above. Maybe we should discuss this in person later..

**Review by HDinger — CHANGES_REQUESTED** [review]
@ingamai Good job 👍 I like the smooth animation on mobile.
I know I made a lot of comments but most of them are really small issues. There are comments about the styling breakpoints. I know that the style guide provides them, but I would still be careful. If you want to, we can discuss this later.

File: app/assets/stylesheets/content/_progress_bar.sass (hunk 3 above)
Comment (HDinger) [inline app/assets/stylesheets/content/_progress_bar.sass:52]: My suggestion would be to remove this from this PR and handle all bugs regarding the roadmap in a separate PR.. There are a lot more issues on this page, especially on mobile..

e.g. a WP with a very long name
bildschirmfoto 2018-06-29 um 14 18 25

**Comment by HDinger** [comment]
It would be nice if the commits could be squashed at the end.

**Comment by HDinger** [comment]
Closing in favor of https://github.com/opf/openproject/pull/6488
